voila mode

bcwu opened the pull request:
### Description

Add support for Voila
- [x] write-manifest voila
    - [x] by file
    - [x] by directory
- [x] deploy manifest voila
- [x] deploy voila
    - [x] by file
    - [x] by directory
- [x] voila.json
- [x] multi-notebook mode

fixes [#22006](https://github.com/rstudio/connect/issues/22006)
fixes [#22049](https://github.com/rstudio/connect/issues/22049)
fixes [#22047](https://github.com/rstudio/connect/issues/22047)

kgartland-rstudio commented:
Issues placeholder...

kgartland-rstudio commented:
deploying voila with no requirements.txt file results in an error

It seems when deploying voila content we aren't running `pip freeze` when the `requirements.txt` file is missing.  This results in Connect failing to build the content and throws the following error:

```
2022/08/29 14:07:32.507057936 Unexpected error while building Python environment: [Errno 2] No such file or directory: 'requirements.txt'
2022/08/29 14:07:32.509883720 Traceback (most recent call last):
2022/08/29 14:07:32.509892493   File "/connect/python/build_environment.py", line 409, in main
2022/08/29 14:07:32.509941456     specs = pip_read_requirements_file(requirements_file)
2022/08/29 14:07:32.509942992   File "/connect/python/report_helpers.py", line 255, in pip_read_requirements_file
2022/08/29 14:07:32.509961389     with open(environment_file, "r") as f:
2022/08/29 14:07:32.509962621 FileNotFoundError: [Errno 2] No such file or directory: 'requirements.txt'
Build error: exit status 255
An error occurred while building the content (build-failed-error)
Error from Connect server: exit status 255
Error: Task exited with status 1.
```

kgartland-rstudio commented:
When deploying voila from a directory where no voila content exists an internal error is thrown

In this example, I was mistakenly in the parent directory which contained only two folders and no voila content.  When I deployed I hit the following error:

```
> rsconnect deploy voila ./ -n yeti-password
    Warning: Capturing the environment using 'pip freeze'.
             Consider creating a requirements.txt file instead.
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: expected str, bytes or os.PathLike object, not NoneType
Traceback (most recent call last):
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/site-packages/rsconnect/main.py", line 86, in wrapper
    result = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/site-packages/rsconnect/main.py", line 811, in deploy_voila
    ce.make_bundle(
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/site-packages/rsconnect/log.py", line 186, in wrapper
    result = method(self, *args, **kw)
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/site-packages/rsconnect/api.py", line 566, in make_bundle
    bundle = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/site-packages/rsconnect/bundle.py", line 849, in make_voila_bundle
    base_dir = dirname(entrypoint)
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/posixpath.py", line 152, in dirname
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType
Internal error: expected str, bytes or os.PathLike object, not NoneType
```

kgartland-rstudio commented:
inaccurate error text when re-deploying a mis-matched content type

re-deploying voila content where a pre-existing `rsconnect-python` folder contains a json file with a non-voila app_mode, results in an error mentioning Python Shiny Application:

```
> rsconnect deploy voila ./ -n yeti-password
    Warning: Capturing the environment using 'pip freeze'.
             Consider creating a requirements.txt file instead.
Validating server... 	[OK]
Validating app mode... 	[ERROR]: Deploying with mode 'Python Shiny Application',
but the existing deployment has mode 'Bokeh Application'.
Use the --new option to create a new deployment of the desired type.
Error: Deploying with mode 'Python Shiny Application',
but the existing deployment has mode 'Bokeh Application'.
Use the --new option to create a new deployment of the desired type.
```
content of the rsconnect-python/*.json file:
```
{
    "https://connect.localtest.me/rsc/dev-password": {
        "server_url": "https://connect.localtest.me/rsc/dev-password",
        "filename": "/Users/kgartland/Projects/voila",
        "app_url": "https://connect.localtest.me:443/rsc/dev-password/content/bb4d3939-86ef-4972-8297-4a81a94c4ca9/",
        "app_id": 19,
        "app_guid": "bb4d3939-86ef-4972-8297-4a81a94c4ca9",
        "title": "no-req",
        "app_mode": "python-bokeh"
    }
```

bcwu commented:
> deploying voila with no requirements.txt file results in an error
> 
> It seems when deploying voila content we aren't running `pip freeze` when the `requirements.txt` file is missing. This results in Connect failing to build the content and throws the following error:
> 
> ```
> 2022/08/29 14:07:32.507057936 Unexpected error while building Python environment: [Errno 2] No such file or directory: 'requirements.txt'
> 2022/08/29 14:07:32.509883720 Traceback (most recent call last):
> 2022/08/29 14:07:32.509892493   File "/connect/python/build_environment.py", line 409, in main
> 2022/08/29 14:07:32.509941456     specs = pip_read_requirements_file(requirements_file)
> 2022/08/29 14:07:32.509942992   File "/connect/python/report_helpers.py", line 255, in pip_read_requirements_file
> 2022/08/29 14:07:32.509961389     with open(environment_file, "r") as f:
> 2022/08/29 14:07:32.509962621 FileNotFoundError: [Errno 2] No such file or directory: 'requirements.txt'
> Build error: exit status 255
> An error occurred while building the content (build-failed-error)
> Error from Connect server: exit status 255
> Error: Task exited with status 1.
> ```

fixed by [564e82a](https://github.com/rstudio/rsconnect-python/pull/277/commits/564e82a7e4de5d5cb4a97ddcf4167a355bb7ae86)



> When deploying voila from a directory where no voila content exists an internal error is thrown
> 
> In this example, I was mistakenly in the parent directory which contained only two folders and no voila content. When I deployed I hit the following error:
> 
> ```
> > rsconnect deploy voila ./ -n yeti-password
>     Warning: Capturing the environment using 'pip freeze'.
>              Consider creating a requirements.txt file instead.
> Validating server... 	[OK]
> Validating app mode... 	[OK]
> Making bundle ... 	[ERROR]: expected str, bytes or os.PathLike object, not NoneType
> Traceback (most recent call last):
>   File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/site-packages/rsconnect/main.py", line 86, in wrapper
>     result = func(*args, **kwargs)
>   File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/site-packages/rsconnect/main.py", line 811, in deploy_voila
>     ce.make_bundle(
>   File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/site-packages/rsconnect/log.py", line 186, in wrapper
>     result = method(self, *args, **kw)
>   File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/site-packages/rsconnect/api.py", line 566, in make_bundle
>     bundle = func(*args, **kwargs)
>   File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/site-packages/rsconnect/bundle.py", line 849, in make_voila_bundle
>     base_dir = dirname(entrypoint)
>   File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/posixpath.py", line 152, in dirname
>     p = os.fspath(p)
> TypeError: expected str, bytes or os.PathLike object, not NoneType
> Internal error: expected str, bytes or os.PathLike object, not NoneType
> ```

fixed by [6fdc811](https://github.com/rstudio/rsconnect-python/pull/277/commits/6fdc81126a1c3a70c7dfd182f8471cfb54f91696)

kgartland-rstudio commented:
The deploy voila help page mentions `static` mode which isn't available in voila deployments:

> rsconnect deploy voila --help
> Usage: rsconnect deploy voila [OPTIONS] PATH [EXTRA_FILES]...
> 
>   Deploy a Voila notebook to RStudio Connect. This may be done by source or as
>   a static HTML page. If the notebook is deployed as a static HTML page
>   (--static), it cannot be scheduled or rerun on the Connect server.

I think we can get rid of the second paragraph and maybe add something about being able to deploy with customizations by adding a `voila.json` in the deploy directory.

kgartland-rstudio commented:
rsconnect version: 1.10.1.dev19+g26be70b

rsconnect write-manifest requires a filename when it shouldn't.  The following command should be able to write a manifest, but instead we're looking for a filename.

>  rsconnect write-manifest voila ./
> Usage: rsconnect write-manifest voila [OPTIONS] FILE [EXTRA_FILES]...
> Try 'rsconnect write-manifest voila --help' for help.
> 
> Error: Invalid value for 'FILE': File './' is a directory.
>

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -760,6 +765,118 @@ def make_html_bundle(
     return bundle_file
 
 
+def pack_extra_files(
```
This function is very similar to make_html_bundle_content. Can we factor out the large common section near the end of the function?

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -725,7 +725,12 @@ def infer_entrypoint(path, mimetype):
-    return mimetype_filelist[mimetype].pop() if len(mimetype_filelist[mimetype]) == 1 else None
+    res = mimetype_filelist[mimetype].pop() if len(mimetype_filelist[mimetype]) == 1 else None
+    if not res:
+        raise RuntimeError(
```
We should consider raising an RSConnectException here. `RuntimeError` is typically raised by the Python language runtime.

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -1266,6 +1383,52 @@ def write_notebook_manifest_json(
     return exists(join(directory, environment.filename))
 
 
+def write_voila_manifest_json(
```
This function currently supports the file-based workflow, like `write-manifest notebook`. Can it be extended to also support the directory-based workflow we have for applications, if the target provided is a directory? Probably better in a separate PR.

mmarchetti commented on rsconnect/main.py:
```diff
@@ -733,6 +735,91 @@ def deploy_notebook(
+# noinspection SpellCheckingInspection,DuplicatedCode
+@deploy.command(
+    name="voila",
+    short_help="Deploy Voila notebook to RStudio Connect [v2022.09.0+].",
```
We should take a note (on our board) to revisit this version number, depending on when the feature lands.

mmarchetti commented on rsconnect/main.py:
```diff
@@ -733,6 +735,91 @@ def deploy_notebook(
+    name="voila",
+    short_help="Deploy Voila notebook to RStudio Connect [v2022.09.0+].",
+    help=(
+        "Deploy a Voila notebook to RStudio Connect. This may be done by source or as a static HTML "
```
The notes about static HTML format don't apply to Voila mode.

mmarchetti commented on rsconnect/main.py:
```diff
@@ -1247,6 +1334,94 @@ def write_manifest_notebook(
+    + "The Python environment must have the rsconnect package installed.",
+)
+@click.option(
+    "--conda",
```
This option should be removed.

mmarchetti commented on rsconnect/main.py:
```diff
@@ -1247,6 +1334,94 @@ def write_manifest_notebook(
+    help="Target image to be used during content execution (only applicable if the RStudio Connect "
+    "server is configured to use off-host execution)",
+)
+@click.argument("file", type=click.Path(exists=True, dir_okay=False, file_okay=True))
```
Are we going to extend this to directory deployment? Maybe in a separate PR would be better.

kgartland-rstudio commented:
Invalid themes cause the notebook to fail to render.

I saw in the example that there's a `jupyterlab_miami_nights` theme so I added that to my voila.json file.  The theme failed and it caused the notebook to fail to render entirely.  Rather than giving the user an error in the Dashboard, a blank page was displayed.  The following error was displayed in the application logs.  

When using an invalid theme, can we display this error in the dashboard?

> ValueError: Could not find lab theme "jupyterlab_miami_nights"

Full Error:

> 2022/09/08 1:47:05 PM: 404 POST /voila/api/shutdown/f2ab7c9e-1a24-40d8-bbc0-23fdc7662133 (127.0.0.1): Kernel does not exist: f2ab7c9e-1a24-40d8-bbc0-23fdc7662133
> 2022/09/08 1:47:05 PM: 404 POST /voila/api/shutdown/f2ab7c9e-1a24-40d8-bbc0-23fdc7662133 (127.0.0.1) 2.90ms
> 2022/09/08 1:47:06 PM: Task exception was never retrieved
> 2022/09/08 1:47:06 PM: future: .put_html() done, defined at /connect/mnt/app/python/env/lib/python3.10/site-packages/voila/handler.py:202> exception=ValueError('Could not find lab theme "jupyterlab_miami_nights"')>
> 2022/09/08 1:47:06 PM: Traceback (most recent call last):
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/lib/python3.10/site-packages/voila/handler.py", line 203, in put_html
> 2022/09/08 1:47:06 PM: async for html_snippet, _ in gen.generate_content_generator(kernel_id, kernel_future):
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/lib/python3.10/site-packages/voila/exporter.py", line 101, in generate_from_notebook_node
> 2022/09/08 1:47:06 PM: async for output in self.template.generate_async(nb=nb_copy, resources=resources, **extra_context, static_url=self.static_url):
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/lib/python3.10/site-packages/jinja2/environment.py", line 1373, in generate_async
> 2022/09/08 1:47:06 PM: yield self.environment.handle_exception()
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
> 2022/09/08 1:47:06 PM: raise rewrite_traceback_stack(source=source)
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/share/jupyter/voila/templates/lab/index.html.j2", line 4, in top-level template code
> 2022/09/08 1:47:06 PM: {% from 'voila_setup.macro.html.j2' import voila_setup_helper_functions, voila_setup_nbextensions with context %}
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/share/jupyter/nbconvert/templates/lab/index.html.j2", line 3, in top-level template code
> 2022/09/08 1:47:06 PM: {% from 'jupyter_widgets.html.j2' import jupyter_widgets %}
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/share/jupyter/nbconvert/templates/lab/base.html.j2", line 280, in top-level template code
> 2022/09/08 1:47:06 PM: {% set div_id = uuid4() %}
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/share/jupyter/nbconvert/templates/base/display_priority.j2", line 1, in top-level template code
> 2022/09/08 1:47:06 PM: {%- extends 'base/null.j2' -%}
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/share/jupyter/nbconvert/templates/base/null.j2", line 24, in top-level template code
> 2022/09/08 1:47:06 PM: {%- block header -%}
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/share/jupyter/nbconvert/templates/lab/index.html.j2", line 9, in block 'header'
> 2022/09/08 1:47:06 PM: {%- block html_head -%}
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/share/jupyter/nbconvert/templates/lab/index.html.j2", line 36, in block 'html_head'
> 2022/09/08 1:47:06 PM: {% block notebook_css %}
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/share/jupyter/nbconvert/templates/lab/index.html.j2", line 43, in block 'notebook_css'
> 2022/09/08 1:47:06 PM: {{ resources.include_lab_theme(resources.theme) }}
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/lib/python3.10/site-packages/voila/utils.py", line 159, in include_lab_theme
> 2022/09/08 1:47:06 PM: theme_name, _ = find_lab_theme(name)
> 2022/09/08 1:47:06 PM: File "/connect/mnt/app/python/env/lib/python3.10/site-packages/nbconvert/exporters/html.py", line 73, in find_lab_theme
> 2022/09/08 1:47:06 PM: raise ValueError(f'Could not find lab theme "{theme_name}"')
> 2022/09/08 1:47:06 PM: ValueError: Could not find lab theme "jupyterlab_miami_nights"

bcwu commented:
> The deploy voila help page mentions `static` mode which isn't available in voila deployments:
> 
> > rsconnect deploy voila --help
> > Usage: rsconnect deploy voila [OPTIONS] PATH [EXTRA_FILES]...
> > Deploy a Voila notebook to RStudio Connect. This may be done by source or as
> > a static HTML page. If the notebook is deployed as a static HTML page
> > (--static), it cannot be scheduled or rerun on the Connect server.
> 
> I think we can get rid of the second sentence and maybe add something about being able to deploy with customizations by adding a `voila.json` in the deploy directory.


fixed by [8725e9f](https://github.com/rstudio/rsconnect-python/pull/277/commits/8725e9fe8486d6743177c47dce5a881c03b33f09)



> rsconnect version: 1.10.1.dev19+g26be70b
> 
> rsconnect write-manifest requires a filename when it shouldn't. The following command should be able to write a manifest, but instead we're looking for a filename.
> 
> > rsconnect write-manifest voila ./
> > Usage: rsconnect write-manifest voila [OPTIONS] FILE [EXTRA_FILES]...
> > Try 'rsconnect write-manifest voila --help' for help.
> > Error: Invalid value for 'FILE': File './' is a directory.

fixed by [e0cfe36](https://github.com/rstudio/rsconnect-python/pull/277/commits/e0cfe36dda411e46f4189ae196b58ff474befca5)

kgartland-rstudio commented:
Logging this as a reminder to update.

```
rsconnect deploy voila --help
  ...
voila          Deploy Voila notebook to RStudio Connect [v2022.09.0+].
```
The help shows the Connect version required is 2022.09.0, we'll need to update that with whatever version it ships with:

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -747,6 +748,69 @@ def make_html_bundle(
+    :param image: the optional docker image to be specified for off-host execution. Default = None.
+    :return: a file-like object containing the bundle tarball.
+    """
+    mimetypes.add_type("text/ipynb", ".ipynb")
```
I'm not sure it matters for how we are using this, but the mimetype for ipynb files appears to be [application/x-ipynb+json](https://github.com/jupyter/notebook/pull/6010/files#diff-b34aa95c7163adbb4c323d5930b11fbbc28ff11984f6436fb139358f7ba5b719)

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -1241,6 +1305,59 @@ def write_notebook_manifest_json(
+    :param image: the optional docker image to be specified for off-host execution. Default = None.
+    :return: whether the manifest was written.
+    """
+    mimetypes.add_type("text/ipynb", ".ipynb")
```
This mimetype should be a constant - it appears several times in the code as a literal.

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -1241,6 +1305,59 @@ def write_notebook_manifest_json(
+    excludes.extend(["manifest.json"])
+    file_list = create_file_list(path, extra_files, excludes)
+    for rel_path in file_list:
+        manifest_add_file(manifest_data, rel_path, path)
```
Up to this point, this function appears identical to `make_voila_bundle`, and we need their behaviors to be identical so that deploying now or writing a manifest for later produce the same bundle. Can you factor out the common section so there is only one version of it?

bcwu replied:
To make this happen I factored out data structures for Manifest and Bundle. This made data passing between functions easier and let us avoid the problem of doing unnecessary IO operations for disparate manifest and bundle related function calls.

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -50,6 +51,118 @@
+mimetypes.add_type("text/ipynb", ".ipynb")
+
+
+class Manifest:
```
I like the new `Manifest` and `Bundle` classes. Let's follow up after this PR and use them for the other deployment types.

mmarchetti commented:
I get this error when trying to deploy a file that is not in the current directory:
```
$ rsconnect deploy voila -n devpw ./notebooks3.8/basics.ipynb
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: Unable to include the file ./notebooks3.8/./notebooks3.8/basics.ipynb in the bundle: No such file or directory
Error: Unable to include the file ./notebooks3.8/./notebooks3.8/basics.ipynb in the bundle: No such file or directory
```

It deploys successfully if I run
```
cd notebooks3.8
rsconnect deploy voila -n devpw ./basics.ipynb
```

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -938,6 +1106,26 @@ def validate_extra_files(directory, extra_files):
+        base_dir = path if isdir(path) else dirname(path)
+        for extra in extra_files:
+            if Path(extra).parent != Path(base_dir):
+                raise RSConnectException(f"{extra} must be under {base_dir}.")
```
Also, this check seems strict. Is it OK to add an extra file that is under a subdirectory? Maybe we need to check that the normalized/absolute `base_dir` is a prefix of the normalized/absolute `extra` path?

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -1245,6 +1433,89 @@ def write_notebook_manifest_json(
+    :return: whether the manifest was written.
+    """
+    manifest = create_voila_manifest(**locals())
+    entrypoint = entrypoint or infer_entrypoint(path=path, mimetype="text/ipynb")
```
Could the manifest entrypoint be used here?

bcwu replied:
The manifest can change (make all files and metadata relative to the entrypoint) so using the original path is safer here.

mmarchetti commented on tests/test_bundle.py:
```diff
@@ -612,7 +612,7 @@ def test_validate_extra_files(self):
         self.assertEqual(
             validate_extra_files(directory, [join(directory, "index.htm")]),
-            ["index.htm"],
+            [os.path.join(directory, "index.htm")],
```
This seems like a duplicate of the first value in the list.

bcwu commented on rsconnect/bundle.py:
```diff
@@ -746,6 +866,54 @@ def make_html_bundle(
+
+    manifest = create_voila_manifest(**locals())
+    if manifest.data.get("files") is None:
+        return None
```
Possible if the function is called directly (with no entrypoint).

mmarchetti replied:
OK. We should consider raising an exception in that case.

bcwu commented:
> I get this error when trying to deploy a file that is not in the current directory:
> 
> ```
> $ rsconnect deploy voila -n devpw ./notebooks3.8/basics.ipynb
> Validating server... 	[OK]
> Validating app mode... 	[OK]
> Making bundle ... 	[ERROR]: Unable to include the file ./notebooks3.8/./notebooks3.8/basics.ipynb in the bundle: No such file or directory
> Error: Unable to include the file ./notebooks3.8/./notebooks3.8/basics.ipynb in the bundle: No such file or directory
> ```
> 
> It deploys successfully if I run
> 
> ```
> cd notebooks3.8
> rsconnect deploy voila -n devpw ./basics.ipynb
> ```

fixed by [69115ae](https://github.com/rstudio/rsconnect-python/pull/277/commits/69115aea4eabeaabc9fd9514649b5771d92e23da)

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -1245,6 +1439,90 @@ def write_notebook_manifest_json(
+
+    # handle environment files
+    if not exists(join(deploy_dir, environment.filename)) or force_generate:
+        manifest.add_file(join(deploy_dir, environment.filename))
```
We always need to add a `requirements.txt` file to the manifest file list, so this can't be conditional.

`environment.contents` contains the requirements file content, either from `requirements.txt` or `pip freeze`. You may need to create Manifest and Bundle methods to add a file buffer, as is done for the other content types, so that you can inject a requirements file into the manifest and bundle without writing it to the directory.

bcwu replied:
fixed by [fcfbdd0](https://github.com/rstudio/rsconnect-python/pull/277/commits/fcfbdd0bbc418a7a881378981a797c579a8cd88a)

bcwu replied:
now that I think about it, the original if condition is correct, because it prioritizes the user provided requirements.txt. Without it, we will always use the system generated one in environment.contents. I reverted the change in [cad1ae6](https://github.com/rstudio/rsconnect-python/pull/277/commits/cad1ae62a07b2b6326b79066bdf7d92044246c8a)

mmarchetti replied:
We need to make sure these cases all work correctly:
* write-manifest with an existing requirements.txt
* write-manifest without requirements.txt
* deploy with an existing requirements.txt
* deploy without requirements.txt

I think that right now, the last one will not work because it won't include a requirements.txt in the bundle. You can unconditionally call
```
manifest.add_to_buffer(environment.filename, environment.contents)
```
because the environment inspection already prioritizes requirements.txt and will return its contents, if the file exists. It only invokes `pip freeze` if there is no requirements file.

kgartland-rstudio commented:
`write-manifest` when targeting a file fails. 

Running this command on the branch here:
`rsconnect write-manifest voila ./index.ipynb`

The following error is thrown. Is `write-manifest` with a file supported, or is it supposed to only take a directory?

```
Checking arguments...                            [OK]
Inspecting Python environment...                 [OK]
    Warning: requirements.txt already exists and will not be overwritten.
Creating manifest.json...                        Traceback (most recent call last):
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/actions.py", line 86, in cli_feedback
    yield
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 1525, in write_manifest_voila
    write_voila_manifest_json(
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 1569, in write_voila_manifest_json
    manifest = create_voila_manifest(**locals())
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 1524, in create_voila_manifest
    excludes.extend(list_environment_dirs(path))
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 747, in list_environment_dirs
    for name in os.listdir(directory):
NotADirectoryError: [Errno 20] Not a directory: './index.ipynb'
[ERROR]
Internal error: [Errno 20] Not a directory: './index.ipynb'
```

kgartland-rstudio commented:
The `--exclude` flag doesn't exist in the deploy voila command.

```
rsconnect deploy voila ./country-indicators/ -n yeti-ldap --exclude country-indicators/extra-folder/
Usage: rsconnect deploy voila [OPTIONS] PATH [EXTRA_FILES]...
Try 'rsconnect deploy voila --help' for help.

Error: No such option: --exclude Did you mean --insecure?
```

kgartland-rstudio commented:
Authorization: Key is exposed in `--verbose` mode. I just noticed this when deploying `notebook` too.  It is not exposed when deploy api or dash.

> Completed Python build against Python version: '3.8.12'
> Launching Voila interactive notebook...
> [DEBUG] 2023-02-14T11:43:35-0500 Request: GET /rsc/dev-ldap/__api__/applications/19/config
> [DEBUG] 2023-02-14T11:43:35-0500 Headers:
> [DEBUG] 2023-02-14T11:43:35-0500 --> User-Agent: rsconnect-python/1.14.2.dev136+g4f6d7ae
> [DEBUG] 2023-02-14T11:43:35-0500 --> Authorization: Key ZI9HQQ1rzre1IX5uxzhMA6WevZw8ZLn1
> [DEBUG] 2023-02-14T11:43:35-0500 Response: 200 OK
> [DEBUG] 2023-02-14T11:43:35-0500 Headers:
> [DEBUG] 2023-02-14T11:43:35-0500 --> Server: nginx/1.19.10
> [DEBUG] 2023-02-14T11:43:35-0500 --> Date: Tue, 14 Feb 2023 16:43:35 GMT
> [DEBUG] 2023-02-14T11:43:35-0500 --> Content-Type: application/json; charset=utf-8
> [DEBUG] 2023-02-14T11:43:35-0500 --> Content-Length: 230
> [DEBUG] 2023-02-14T11:43:35-0500 --> Connection: keep-alive
> [DEBUG] 2023-02-14T11:43:35-0500 --> Cache-Control: no-cache, no-store, must-revalidate
> [DEBUG] 2023-02-14T11:43:35-0500 --> Expires: 0
> [DEBUG] 2023-02-14T11:43:35-0500 --> Pragma: no-cache
> [DEBUG] 2023-02-14T11:43:35-0500 --> Strict-Transport-Security: max-age=2592000
> [DEBUG] 2023-02-14T11:43:35-0500 --> X-Content-Type-Options: nosniff
> [DEBUG] 2023-02-14T11:43:35-0500 --> X-Correlation-Id: 20a009b7-13fb-4b0f-b13d-0589a1628d64
> [DEBUG] 2023-02-14T11:43:35-0500 --> {"config_url":"https://connect.localtest.me:443/rsc/dev-ldap/connect/#/apps/5ce9c0c4-2e97-4def-bf1d-c5b1e51d7d99","logs_url":"https://connect.localtest.me:443/rsc/dev-ldap/connect/#/apps/5ce9c0c4-2e97-4def-bf1d-c5b1e51d7d99/logs"}
> [DEBUG] 2023-02-14T11:43:35-0500 CookieJar contents: []
> {}
> Deployment completed successfully.

bcwu commented:
> Authorization: Key is exposed in `--verbose` mode. I just noticed this when deploying `notebook` too. It is not exposed when deploy api or dash.
> 
> > Completed Python build against Python version: '3.8.12'
> > Launching Voila interactive notebook...
> > [DEBUG] 2023-02-14T11:43:35-0500 Request: GET /rsc/dev-ldap/**api**/applications/19/config
> > [DEBUG] 2023-02-14T11:43:35-0500 Headers:
> > [DEBUG] 2023-02-14T11:43:35-0500 --> User-Agent: rsconnect-python/1.14.2.dev136+g4f6d7ae
> > [DEBUG] 2023-02-14T11:43:35-0500 --> Authorization: Key ZI9HQQ1rzre1IX5uxzhMA6WevZw8ZLn1
> > [DEBUG] 2023-02-14T11:43:35-0500 Response: 200 OK
> > [DEBUG] 2023-02-14T11:43:35-0500 Headers:
> > [DEBUG] 2023-02-14T11:43:35-0500 --> Server: nginx/1.19.10
> > [DEBUG] 2023-02-14T11:43:35-0500 --> Date: Tue, 14 Feb 2023 16:43:35 GMT
> > [DEBUG] 2023-02-14T11:43:35-0500 --> Content-Type: application/json; charset=utf-8
> > [DEBUG] 2023-02-14T11:43:35-0500 --> Content-Length: 230
> > [DEBUG] 2023-02-14T11:43:35-0500 --> Connection: keep-alive
> > [DEBUG] 2023-02-14T11:43:35-0500 --> Cache-Control: no-cache, no-store, must-revalidate
> > [DEBUG] 2023-02-14T11:43:35-0500 --> Expires: 0
> > [DEBUG] 2023-02-14T11:43:35-0500 --> Pragma: no-cache
> > [DEBUG] 2023-02-14T11:43:35-0500 --> Strict-Transport-Security: max-age=2592000
> > [DEBUG] 2023-02-14T11:43:35-0500 --> X-Content-Type-Options: nosniff
> > [DEBUG] 2023-02-14T11:43:35-0500 --> X-Correlation-Id: 20a009b7-13fb-4b0f-b13d-0589a1628d64
> > [DEBUG] 2023-02-14T11:43:35-0500 --> {"config_url":"[https://connect.localtest.me:443/rsc/dev-ldap/connect/#/apps/5ce9c0c4-2e97-4def-bf1d-c5b1e51d7d99","logs_url":"https://connect.localtest.me:443/rsc/dev-ldap/connect/#/apps/5ce9c0c4-2e97-4def-bf1d-c5b1e51d7d99/logs"}](https://connect.localtest.me:443/rsc/dev-ldap/connect/#/apps/5ce9c0c4-2e97-4def-bf1d-c5b1e51d7d99%22,%22logs_url%22:%22https://connect.localtest.me:443/rsc/dev-ldap/connect/#/apps/5ce9c0c4-2e97-4def-bf1d-c5b1e51d7d99/logs%22%7D)
> > [DEBUG] 2023-02-14T11:43:35-0500 CookieJar contents: []
> > {}
> > Deployment completed successfully.

Let's file this as a separate issue. When I added set_verbosity(verbose) to deploy_api, it started exposing the key as well.

bcwu commented:
> The `--exclude` flag doesn't exist in the deploy voila command.
> 
> ```
> rsconnect deploy voila ./country-indicators/ -n yeti-ldap --exclude country-indicators/extra-folder/
> Usage: rsconnect deploy voila [OPTIONS] PATH [EXTRA_FILES]...
> Try 'rsconnect deploy voila --help' for help.
> 
> Error: No such option: --exclude Did you mean --insecure?
> ```

fixed by [ce06ff6](https://github.com/rstudio/rsconnect-python/pull/277/commits/ce06ff63bb7535a8dfeebc975601378240b077e9)

bcwu commented:
> `write-manifest` when targeting a file fails.
> 
> Running this command on the branch here: `rsconnect write-manifest voila ./index.ipynb`
> 
> The following error is thrown. Is `write-manifest` with a file supported, or is it supposed to only take a directory?
> 
> ```
> Checking arguments...                            [OK]
> Inspecting Python environment...                 [OK]
>     Warning: requirements.txt already exists and will not be overwritten.
> Creating manifest.json...                        Traceback (most recent call last):
>   File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/actions.py", line 86, in cli_feedback
>     yield
>   File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 1525, in write_manifest_voila
>     write_voila_manifest_json(
>   File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 1569, in write_voila_manifest_json
>     manifest = create_voila_manifest(**locals())
>   File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 1524, in create_voila_manifest
>     excludes.extend(list_environment_dirs(path))
>   File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 747, in list_environment_dirs
>     for name in os.listdir(directory):
> NotADirectoryError: [Errno 20] Not a directory: './index.ipynb'
> [ERROR]
> Internal error: [Errno 20] Not a directory: './index.ipynb'
> ```



> `write-manifest` from outside the directory will create a `requirements.txt` file in the current directory rather than the target directory unless a trailing `/` is included in the directory argument.
> 
> `rsconnect write-manifest voila country-indicators` or `rsconnect write-manifest voila ./country-indicators` will create a requirements file in the current working directory
> 
> `rsconnect write-manifest voila country-indicators/` or `rsconnect write-manifest voila ./country-indicators/` will create a requirements file in the target directory (`country-indicators`).
> 
> I tested with `dash` and `api` and the requirements.txt file was correctly created in the target directory for those content types.

both should be fixed by [b3bcef6](https://github.com/rstudio/rsconnect-python/pull/277/commits/b3bcef63ba4859c680acef62503e2f80f0816912)

kgartland-rstudio commented:
The `deploy voila` command now supports the `--exclude` argument but does not exclude sub-folders like other content-types do.

i.e.
`rsconnect deploy voila ./country-indicators/ --exclude country-indicators/extra-folder/ -n yeti-ldap`

```
country-indicators/
├── __pycache__
├── env
├── extra-file.txt
├── extra-folder
│   ├── files.txt
│   ├── subfolder
│   │   └── subfile.txt
```

results in the `extra-folder/subfolder/subfile.txt` to be included in the bundle.

I would expect all contents of the `extra-folder` directory to be excluded in this deployment.  I verified with dash deploys, we don't include subfolders of the excluded directory.

bcwu commented:
> The `deploy voila` command now supports the `--exclude` argument but does not exclude sub-folders like other content-types do.
> 
> i.e. `rsconnect deploy voila ./country-indicators/ --exclude country-indicators/extra-folder/ -n yeti-ldap`
> 
> ```
> country-indicators/
> ├── __pycache__
> ├── env
> ├── extra-file.txt
> ├── extra-folder
> │   ├── files.txt
> │   ├── subfolder
> │   │   └── subfile.txt
> ```
> 
> results in the `extra-folder/subfolder/subfile.txt` to be included in the bundle. It does exclude the files in the immediate excluded directory but not subfolders.
> 
> I would expect all contents of the `extra-folder` directory to be excluded in this deployment. I verified with dash deploys, we don't include subfolders of the excluded directory.

fixed by [69cf3f0](https://github.com/rstudio/rsconnect-python/pull/277/commits/69cf3f0cd7b0839d04abad8aab669a05cac2bc13)

We should open a separate issue about glob set not working. E.g. if I have ./folder1/folder2/ignoreme.txt and I run `deploy [any mode] ./ --exclude ignoreme.txt`, ignoreme.txt won't get excluded because it's nested and not relative to the deploy directory.

kgartland-rstudio commented:
> We should open a separate issue about glob set not working. E.g. if I have ./folder1/folder2/ignoreme.txt and I run `deploy [any mode] ./ --exclude ignoreme.txt`, ignoreme.txt won't get excluded because it's nested and not relative to the deploy directory.

That behavior sounds right to me. You should have to `--exclude ./folder1/folder2/ignoreme.txt` to exclude that file. @mmarchetti thoughts?

kgartland-rstudio commented:
Related to: 
https://github.com/rstudio/rsconnect-python/pull/277#issuecomment-1230388112

In interest of testing multi-notebook support, I also attempted this where I have a number of folders containing various notebooks and hit the same error.  

```
├── country-indicators
│   ├── index.ipynb
├── cpp-xleaflet
│   ├── README.md
│   ├── index.ipynb
│   ├── jupyter_config.json
│   └── thumbnail.png
├── fileupload
│   ├── app.ipynb
├── gaussian-density
│   ├── LICENSE
│   ├── README.md
│   ├── index.ipynb
│   ├── requirements.txt
```

```> rsconnect deploy voila . -n yeti-pam -N
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: expected str, bytes or os.PathLike object, not NoneType
Traceback (most recent call last):
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 95, in wrapper
    result = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 923, in deploy_voila
    ce.make_bundle(
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/log.py", line 186, in wrapper
    result = method(self, *args, **kw)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/api.py", line 636, in make_bundle
    bundle = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 974, in make_voila_bundle
    deploy_dir = dirname(entrypoint)
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/posixpath.py", line 152, in dirname
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType
Internal error: expected str, bytes or os.PathLike object, not NoneType
```

I then moved two notebooks into the same directory and tried to deploy by directory and hit the same error.  
```
> rsconnect deploy voila . -n yeti-pam -N
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: expected str, bytes or os.PathLike object, not NoneType
Traceback (most recent call last):
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 95, in wrapper
    result = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/main.py", line 923, in deploy_voila
    ce.make_bundle(
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/log.py", line 186, in wrapper
    result = method(self, *args, **kw)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/api.py", line 636, in make_bundle
    bundle = func(*args, **kwargs)
  File "/Users/kgartland/.pyenv/versions/3.8.2/envs/rsconnect-tag/lib/python3.8/site-packages/rsconnect/bundle.py", line 974, in make_voila_bundle
    deploy_dir = dirname(entrypoint)
  File "/Users/kgartland/.pyenv/versions/3.8.2/lib/python3.8/posixpath.py", line 152, in dirname
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType
Internal error: expected str, bytes or os.PathLike object, not NoneType
```

Further deploying by filename while specifying the two notebooks, the first notebook seems to win:

```
> rsconnect deploy voila ./index2.ipynb ./index.ipynb -n yeti-pam -N
```

It seems like we should have a new argument for multi-notebook support if we want to allow voila multi-notebook deployment outside of `deploy manifest`.

mmarchetti commented:
> You should have to `--exclude ./folder1/folder2/ignoreme.txt` to exclude that file

Yes, that sounds right.

bcwu commented:
https://github.com/rstudio/rsconnect-python/pull/277#issuecomment-1230388112
fixed by [cdd27b1](https://github.com/rstudio/rsconnect-python/pull/277/commits/cdd27b1b9964d87d0aa8b31fd037284e7cb4be04)

Instead of showing an internal error, an exception message is raised.

https://github.com/rstudio/rsconnect-python/pull/277#issuecomment-1431909228
fixed by [cdd27b1](https://github.com/rstudio/rsconnect-python/pull/277/commits/cdd27b1b9964d87d0aa8b31fd037284e7cb4be04)

Fixed deploying multi-notebook mode by top directory
An exception message is shown on how to deploy multi-notebook mode if the requirements are not met:

```
                Unable to infer entrypoint from multiple candidates.
                Multi-notebook deployments need to be specified with the following:
                1) An empty string entrypoint, i.e. ""
                2) A directory as the path
```

mmarchetti commented on rsconnect/bundle.py:
```diff
@@ -746,6 +952,84 @@ def make_html_bundle(
+    manifest_flattened_copy_data = manifest.flattened_copy.data
+    if multi_notebook_mode and "metadata" in manifest_flattened_copy_data:
+        manifest_flattened_copy_data["metadata"]["entrypoint"] = ""
+    write_manifest_json(manifest_path, manifest_flattened_copy_data)
```
This creates manifest.json in the deployment directory. We should be creating it in the Bundle but not modifying the contents of the current directory.

bcwu replied:
fixed by [6ba4a38](https://github.com/rstudio/rsconnect-python/pull/277/commits/6ba4a382f6bbe5ea6cc32593e0ee5f026884c86e)

kgartland-rstudio commented:
In Windows you can add a file to the manifest that is not under the target directory.

`rsconnect write-manifest voila .\voila\index.ipynb .\outside-folder\test.txt`

That results in a manifest that will include `.\\outside-folder\\test.txt`.

When running the same command with `notebook`, it throws the proper error:

`rsconnect write-manifest notebook .\voila\index.ipynb .\outside-folder\test.txt`

```
Error: Could not find file ..\outside-folder\test.txt under .\voila
```

bcwu commented:
> In Windows you can add a file to the manifest that is not under the target directory.
> 
> `rsconnect write-manifest voila .\voila\index.ipynb .\outside-folder\test.txt`
> 
> That results in a manifest that will include `.\\outside-folder\\test.txt`.
> 
> When running the same command with `notebook`, it throws the proper error:
> 
> `rsconnect write-manifest notebook .\voila\index.ipynb .\outside-folder\test.txt`
> 
> ```
> Error: Could not find file ..\outside-folder\test.txt under .\voila
> ```

fixed by [e507f10](https://github.com/rstudio/rsconnect-python/pull/277/commits/e507f104af8e5736994020c3d9231410bb7e0092)

kgartland-rstudio commented:
using the new --multi-notebook flag fails if I don't supply an empty string for `--entrypoint`. 

```
> rsconnect deploy voila ./ --multi-notebook -s http://localhost:3939/ -k ivdJSJliii9DZxO2q1N8v0PkStaKKRty
    Warning: Capturing the environment using 'pip freeze'.
             Consider creating a requirements.txt file instead.
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: No valid entrypoint provided or found.
Error: No valid entrypoint provided or found.
```

Should this instead say:
```
                Unable to infer entrypoint from multiple candidates.
                Multi-notebook deployments need to be specified with the following:
                1) An empty string entrypoint, i.e. ""
                2) A directory as the path
```

RE: https://github.com/rstudio/rsconnect-python/pull/277#issuecomment-1433109628

bcwu commented:
> using the new --multi-notebook flag does not hit the expected exception if I don't supply an empty string for `--entrypoint`.
> 
> 
> RE: [#277 (comment)](https://github.com/rstudio/rsconnect-python/pull/277#issuecomment-1433109628)

I see, this is actually an instance of multi-notebook entrypoints nested in subfolders and don't exist in the deploy directory. I relaxed the check in [a35874a](https://github.com/rstudio/rsconnect-python/pull/277/commits/a35874ac2f422bba82f9901b602b70dcf575f3e2). Now we don't check entrypoint when you use `--multi-notebook`; we assume that you have valid notebooks somewhere in the nested directory. 

We don't check entrypoint for empty strings anymore. This is exception that is shown when you deploy a directory with multiple (or no) entrypoints and if you forgot to use  `--multi-notebook`:

```
                Unable to infer entrypoint from multiple candidates.
                Multi-notebook deployments need to be specified with the following:
                1) A directory as the path
                2) Set multi_notebook=True,
                    i.e. include --multi-notebook (or -m) in the CLI command.

```

bcwu commented:
I believe the `--multi-notebook` requirement and various commits solved https://github.com/rstudio/rsconnect-python/pull/277#issuecomment-1431909228

edavidaja commented:
Can we make sure that consistent with the other frameworks, calling `rsconnect deploy voila` returns the help rather than an error?

edavidaja commented:
```
voila/notebooks on  main [!?]
❯ ls
 basics.ipynb      data.csv               interactive.ipynb   jdash.ipynb           mimerenderers.ipynb      reveal.ipynb
 bqplot.ipynb      gridspecLayout.ipynb   ipympl.ipynb        jupyter_config.json   multiple_widgets.ipynb   stock-report-voila.ipynb
 dashboard.ipynb   hash.ipynb             ipyvolume.ipynb     manifest.json         query-strings.ipynb      xleaflet.ipynb

voila/notebooks on  main [!?]
❯ rsconnect deploy voila -n radixu ./jdash.ipynb
    Warning: the existing manifest.json file will not be used or considered.
    Warning: Capturing the environment using 'pip freeze'.
             Consider creating a requirements.txt file instead.
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: Path and entrypoint need to match if they are are files.
Error: Path and entrypoint need to match if they are are files.

voila/notebooks on  main [!?]
❯ rsconnect deploy voila -n radixu --entrypoint ./jdash.ipynb ./jdash.ipynb
    Warning: the existing manifest.json file will not be used or considered.
    Warning: Capturing the environment using 'pip freeze'.
             Consider creating a requirements.txt file instead.
Validating server... 	[OK]
Validating app mode... 	[OK]
Making bundle ... 	[ERROR]: Path and entrypoint need to match if they are are files.
Error: Path and entrypoint need to match if they are are files.

```

bcwu commented:
> Can we make sure that consistent with the other frameworks, calling `rsconnect deploy voila` returns the help rather than an error?

Fixed by [8311a1e](https://github.com/rstudio/rsconnect-python/pull/277/commits/8311a1ea6bf885d946578c4d453c76319835c668), which also fixes https://github.com/rstudio/rsconnect-python/pull/277#issuecomment-1387279363

> ```
> voila/notebooks on  main [!?]
> ❯ ls
>  basics.ipynb      data.csv               interactive.ipynb   jdash.ipynb           mimerenderers.ipynb      reveal.ipynb
>  bqplot.ipynb      gridspecLayout.ipynb   ipympl.ipynb        jupyter_config.json   multiple_widgets.ipynb   stock-report-voila.ipynb
>  dashboard.ipynb   hash.ipynb             ipyvolume.ipynb     manifest.json         query-strings.ipynb      xleaflet.ipynb
> 
> voila/notebooks on  main [!?]
> ❯ rsconnect deploy voila -n radixu ./jdash.ipynb
>     Warning: the existing manifest.json file will not be used or considered.
>     Warning: Capturing the environment using 'pip freeze'.
>              Consider creating a requirements.txt file instead.
> Validating server... 	[OK]
> Validating app mode... 	[OK]
> Making bundle ... 	[ERROR]: Path and entrypoint need to match if they are are files.
> Error: Path and entrypoint need to match if they are are files.
> 
> voila/notebooks on  main [!?]
> ❯ rsconnect deploy voila -n radixu --entrypoint ./jdash.ipynb ./jdash.ipynb
>     Warning: the existing manifest.json file will not be used or considered.
>     Warning: Capturing the environment using 'pip freeze'.
>              Consider creating a requirements.txt file instead.
> Validating server... 	[OK]
> Validating app mode... 	[OK]
> Making bundle ... 	[ERROR]: Path and entrypoint need to match if they are are files.
> Error: Path and entrypoint need to match if they are are files.
> ```

fixed by [0165f23](https://github.com/rstudio/rsconnect-python/pull/277/commits/0165f23100841d8a95f152db08cc1ac37c88ad3e)

kgartland-rstudio commented:
Specifying multiple files using `--multi-notebook` flag fails with a misleading error.

```
> rsconnect deploy voila ./app.ipynb ./index.ipynb --multi-notebook -n local
```

```
 	[ERROR]:
Unable to infer entrypoint.
Multi-notebook deployments need to be specified with the following:
1) A directory as the path
2) Set multi_notebook=True,
    i.e. include --multi-notebook (or -m) in the CLI command.

Error:
Unable to infer entrypoint.
Multi-notebook deployments need to be specified with the following:
1) A directory as the path
2) Set multi_notebook=True,
    i.e. include --multi-notebook (or -m) in the CLI command.
```

Is mutli-notebook mode supported when using file deployment rather than directory deployment?

bcwu commented:
> Specifying multiple files using `--multi-notebook` flag fails with a misleading error.
> 
> ```
> > rsconnect deploy voila ./app.ipynb ./index.ipynb --multi-notebook -n local
> ```
> 
> ```
>  	[ERROR]:
> Unable to infer entrypoint.
> Multi-notebook deployments need to be specified with the following:
> 1) A directory as the path
> 2) Set multi_notebook=True,
>     i.e. include --multi-notebook (or -m) in the CLI command.
> 
> Error:
> Unable to infer entrypoint.
> Multi-notebook deployments need to be specified with the following:
> 1) A directory as the path
> 2) Set multi_notebook=True,
>     i.e. include --multi-notebook (or -m) in the CLI command.
> ```
> 
> Is mutli-notebook mode supported when using file deployment rather than directory deployment?

Multi-notebook mode is not supported using file deployment. It must be directory deployment per requirement 1) in the exception message.

kgartland-rstudio commented:
Voila deployments from rsconnect-jupyter includes all files in the directory by default.  In other notebooks you have to specify which files to include using the Additional Files dialog.

This seems to be publishing as directory rather than by file, is that expected?

Plugin versions:
```
rsconnect-jupyter server extension version: 1.6.1.dev36+g7d00034
rsconnect-jupyter nbextension version: 1.6.1.dev36+g7d00034
rsconnect-python version:1.14.2.dev164+g6a2973a
```

Files deployed when deploying as voila:
```
├── README.md
├── app.ipynb
├── environment.yml
├── manifest.json
├── no-theme
├── none.txt
├── requirements.txt
└── thumbnail.png
```

Files deployed when deploying the same notebook as a jupyter notebook:
```
├── app.ipynb
├── manifest.json
└── requirements.txt
```

bcwu commented:
> Voila deployments from rsconnect-jupyter includes all files in the directory by default. In other notebooks you have to specify which files to include using the Additional Files dialog.
> 
> This seems to be publishing as directory rather than by file, is that expected?
> 
> Plugin versions:
> 
> ```
> rsconnect-jupyter server extension version: 1.6.1.dev36+g7d00034
> rsconnect-jupyter nbextension version: 1.6.1.dev36+g7d00034
> rsconnect-python version:1.14.2.dev164+g6a2973a
> ```
> 
> Files deployed when deploying as voila:
> 
> ```
> ├── README.md
> ├── app.ipynb
> ├── environment.yml
> ├── manifest.json
> ├── no-theme
> ├── none.txt
> ├── requirements.txt
> └── thumbnail.png
> ```
> 
> Files deployed when deploying the same notebook as a jupyter notebook:
> 
> ```
> ├── app.ipynb
> ├── manifest.json
> └── requirements.txt
> ```

Currently in the CLI we make the following distinction: 
- if you want to deploy a directory:
  - you enter the directory as the path
  - whether you enter an entrypoint or not you'll still get the entire folder
- if you want to deploy just a single file:
  - you have to entire that file as the path


@mmarchetti 
Which means the following needs to be changed if we don't want the entire folder:
https://github.com/rstudio/rsconnect-jupyter/blob/7d00034f42721e3225ffbab7faf32d96f0a1e655/rsconnect_jupyter/__init__.py#L205-L206

from
```
dirname(os_path),
basename(os_path),
```

to just:

```
os_path,
None,
```

mmarchetti commented:
I made the change to deploy in single file mode.

kgartland-rstudio commented:
> I made the change to deploy in single file mode.

Verified rsconnect-jupyter now deploys single file mode.
